### ISSUE TITLE
Support for 512x512

### ISSUE THREAD

**Issue by cyberpunkbln** [mdsteele/rust-ico#3]
Hello,
please can you support 512x512 Icons. Decode is not possible for that special format and the problem is most of .ico writes 256x256 metdata in the icondirectory and i can't sniff the nt decodable icons.
thx

**Comment by mdsteele** [comment]
Could you point me to an example of an `.ico` file that this crate is failing to decode?

**Comment by cyberpunkbln** [comment]
@mdsteele 
[ID.zip](https://github.com/mdsteele/rust-ico/files/10252150/ID.zip)
IconDirEntry = 256x256
IconImage = 512x512
I think this is an hack, that old software and old windows don't reject the whole ico. Sniff the first (that was an 256x256 in metadata) and then decode it in 512x512 like binary data.
In the moment i use quirky error handling
```

                for entry in icon_dir.entries()
		{
			let image = entry.decode();
			match &image
			{
				Ok(file) =>
				{
					// image.unwrap();
					let file = std::fs::File::create(exePath.to_owned()+r"\"+exeFile+".png").unwrap();
					image.unwrap().write_png(file).unwrap();
					break;
				},
	        	Err(_) => (),
			}
```
But in fact i have a functioned 512x512 png image in that ico and i must use error handling to stepover.
thx

**Comment by cyberpunkbln** [comment]
please can you make:

```
const MAX_WIDTH: u32 = 256;
const MAX_HEIGHT: u32 = 256;

```

configurable? i have many exes found that have icons above 256, some 2048

**Comment by mdsteele** [comment]
Thanks very much for the file to test on.  I finally found this footnote [on Wikipedia](https://en.wikipedia.org/wiki/ICO_(file_format)#cite_note-bigSize-8):

>  Since Windows 95 size of an image in the ICONDIRENTRY structure might be set to zero which means 256 pixels. Since Windows Vista the size of an image is inferred from the BITMAPFILEHEADER structure or PNG image data which allows storing icons with sizes more than 256.

Apparently I had been working from a long-outdated understanding of the format; looks like nowadays, a width/height byte of zero in the ICONDIR structure no longer means a size of exactly 256, but rather 256 or more, with the exact size coming from the image data.  And indeed that's what your example file seems to be doing.

I'll take a crack at getting this fixed; shouldn't be too hard.

**Comment by mdsteele** [comment]
This should now be fixed by [5cf7a44](https://github.com/mdsteele/rust-ico/commit/5cf7a44396aaf8edb36d0a7d6338dc2770c8c7a1).

**Comment by cyberpunkbln** [comment]
@mdsteele thx

**Comment by mdsteele** [comment]
Published as [v0.3.0](https://crates.io/crates/ico/0.3.0).
